### ISSUE TITLE
Cannot fit concatenated model to two datasets as input

### ISSUE THREAD

**Issue by ch-hristov** [tensorflow/tensorflow#43423]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: 
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.3.0
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.1
- GPU model and memory: MX 150

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior** Can't input two datasets when a model merges two models. The error I get is : 

`Failed to find data adapter that can handle input: (<class 'list'> containing values of types {"<class 'tensorflow.python.data.ops.dataset_ops.TensorDataset'>"}), <class 'NoneType'>`

**Describe the expected behavior** I should be able to input the two data sets or I should be able to somehow merge them together into a single dataset

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

Code : 
```

import tensorflow as tf

# Create first model
model1 = tf.keras.Sequential()
model1.add(tf.keras.layers.Dense(1))
model1.compile()
model1.build([None,3])

# Create second model
model2 = tf.keras.Sequential()
model2.add(tf.keras.layers.Dense(1))
model2.compile()
model2.build([None,3])


# Concatenate
fusion_model = tf.keras.layers.Concatenate()([model1.output, model2.output])
t = tf.keras.layers.Dense(1, activation='tanh')(fusion_model)
model = tf.keras.models.Model(inputs=[model1.input, model2.input], outputs=t)
model.compile()

#Datasets
ds1 = tf.data.Dataset.from_tensors(([1,2,3],1))
ds2 = tf.data.Dataset.from_tensors(([1,2,3], 2))

print(ds1)
print(ds2)
# Fit
model.fit([ds1,ds2])
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by bhack** [comment]
I think you could try to use `tf.data.Dataset.zip`

**Comment by ch-hristov** [comment]
@bhack .zip would try to zip the labels as well. In previous versions you would input the input parameters as [i1,i2,i_n]

**Comment by ch-hristov** [comment]
I also did:

```
 ps = zip(train_text_v, train_ds)

     for i, i2 in ps:
      txt_data = i[0]
      vis_data = i2[0]
      labels = i[1]

      print("INPUT SHAPES!")
      print(txt_data['input_token'].shape)
      print(txt_data['masked_token'].shape)
      print(vis_data.shape)

      yield ({
          'input_token' : txt_data['input_token'], 
          'masked_token' : txt_data['masked_token'], 
          'visual' : vis_data
      }, labels)
```


but that still gives me a problem `as_list() is not defined on an unknown TensorShape`.

**Comment by bhack** [comment]
Can you post you code with the `from_generator`?

**Comment by ch-hristov** [comment]
@bhack here's the full version (including .fit())


```
    def create_early_fusion_ds():
     ps = zip(train_text_v, train_ds)

     for i, i2 in ps:
      txt_data = i[0]
      vis_data = i2[0]
      labels = i[1]

      print("INPUT SHAPES!")
      print(txt_data['input_token'].shape)
      print(txt_data['masked_token'].shape)
      print(vis_data.shape)

      yield {
          'input_token' : txt_data['input_token'], 
          'masked_token' : txt_data['masked_token'], 
          'visual' : vis_data
      }, labels

    ds_train = tf.data.Dataset.from_generator(generator = create_early_fusion_ds,
                                                       output_types=(
                                                               {
                                                                  'input_token' : tf.int64,
                                                                  'masked_token' : tf.int64, 
                                                                  'visual': tf.float64
                                                               }, tf.int64),
                                                        output_shapes = (
                                                            {
                                                                'input_token' : tf.TensorShape([None, 300]),
                                                                'masked_token' : tf.TensorShape([None, 300]),
                                                                'visual' : tf.TensorShape([None, img_width, img_width, 3])
                                                            }, tf.TensorShape([None, len(classes), ])
                                                        )
                                                      )
                    
#     model.save("sxport.h5" , save_format='tf')
    history = model.fit(ds_train,
                        #validation_data = create_early_fusion_ds_val(),
                        epochs = 10,
                        batch_size = 128,
                        callbacks = [es])
```

**Comment by bhack** [comment]
Please re-edit the last post with a very minimal buy complete, copy, paste and runnnable example.
